File: apps/web/src/lib/api/schemas/billing.ts
import { z } from "zod";

import { billingPlanIdSchema, billingTermsVersionSchema, planEntitlementsSchema } from "./pricing";
import { userSchema } from "./users";

const timestampSchema = z.string().datetime({ offset: true });

/**
 * The page to send the customer to.
 *
 * A URL and nothing else. Cards are collected, stored, and shown by the payment
 * provider and never reach this platform, so there is nothing else here to
 * render — the flow is a redirect, not a form.
 */
export const billingHostedSessionResponseSchema = z
  .object({
    url: z.string().min(1),
  })
  .strict();
export type BillingHostedSessionResponse = z.infer<typeof billingHostedSessionResponseSchema>;

export const creditPurchaseSchema = z
  .object({
    id: z.string().uuid(),
    amount_nanos: z.number().int().positive(),
    status: z.enum(["pending", "action_required", "succeeded", "declined", "cancelled"]),
    checkout_url: z.string().nullable(),
    funded_at: timestampSchema.nullable(),
    reversed_nanos: z.number().int().nonnegative(),
  })
  .strict();
export type CreditPurchase = z.infer<typeof creditPurchaseSchema>;

export const creditBalanceSchema = z
  .object({
    ready: z.boolean(),
    balance_nanos: z.number().int(),
  })
  .strict();
export type CreditBalance = z.infer<typeof creditBalanceSchema>;

export const billingPreferencesSchema = z
  .object({
    monthly_usage_limit_nanos: z
      .number()
      .int()
      .nonnegative()
      .max(Number.MAX_SAFE_INTEGER)
      .nullable(),
    reload_enabled: z.boolean(),
    reload_threshold_cents: z.number().int().nonnegative(),
    reload_amount_cents: z.number().int().positive(),
    reload_monthly_payment_limit_cents: z.number().int().nonnegative().nullable(),
  })
  .strict();
export type BillingPreferences = z.infer<typeof billingPreferencesSchema>;

export const automaticReloadStatusSchema = z
  .object({
    paused_purchase_id: z.string().uuid().nullable(),
    pause_reason: z.enum(["declined", "action_required"]).nullable(),
    pending_purchase_id: z.string().uuid().nullable(),
    month_started_at: timestampSchema,
    month_ended_at: timestampSchema,
    monthly_payment_committed_cents: z.number().int().nonnegative(),
  })
  .strict();

export const usageBudgetSchema = z
  .object({
    month_started_at: timestampSchema,
    month_ended_at: timestampSchema,
    limit_nanos: z.number().int().nonnegative().nullable(),
    spent_nanos: z.number().int().nonnegative(),
    available_nanos: z.number().int().nonnegative().nullable(),
  })
  .strict();

export const billingAccountStatuses = ["active", "past_due"] as const;
export type BillingAccountStatus = (typeof billingAccountStatuses)[number];

/** Verified subscription terms and cycle dates, separate from spendable credit. */
export const billingPlanSchema = z
  .object({
    id: billingPlanIdSchema,
    terms_version: billingTermsVersionSchema.nullable(),
    monthly_nanos: z.number().int().nonnegative().nullable(),
    included_nanos: z.number().int().nonnegative().nullable(),
    scheduled_terms_version: billingTermsVersionSchema.nullable(),
    scheduled_change_at: timestampSchema.nullable(),
    // Sent by the server from the same rate card as the pricing endpoint, so the
    // dashboard needs no local map for a plan name.
    name: z.string(),
    period_started_at: timestampSchema.nullable(),
    period_ended_at: timestampSchema.nullable(),
  })
  .strict();
export type BillingPlan = z.infer<typeof billingPlanSchema>;

export const billingEntitlementUsageSchema = z
  .object({
    concurrent_cpu_containers: z.number().int().nonnegative(),
    concurrent_gpus: z.number().int().nonnegative(),
    workspaces: z.number().int().nonnegative(),
    members: z.number().int().nonnegative(),
    connected_clouds: z.number().int().nonnegative(),
    custom_domains: z.number().int().nonnegative(),
  })
  .strict();

export const billingSummarySchema = z
  .object({
    status: z.enum(billingAccountStatuses),
    currency: z.string().regex(/^[A-Z]{3}$/),
    // Null for an account on no subscription. Signing in puts an account on the
    // free plan, so anyone who did has one; what is left is an account reaching
    // the dashboard on a token an administrator minted for it, and one whose
    // subscription the provider says has ended. Neither has terms to be shown, and
    // both need the same thing next, which is a plan.
    plan: billingPlanSchema.nullable(),
    // Whether there is anything at the provider to manage. Signing in registers
    // the customer, so this is true for anyone who did; an account reaching the
    // dashboard on a token an administrator minted for it has no customer record,
    // and the management page would have nobody to show.
    portal_available: z.boolean(),
    // A different question from the one above, which is true from the moment a
    // customer record exists and so cannot tell an account that has saved a card
    // from one that has not. How much the account is given, and whether its
    // running work is stopped when that is spent, both turn on this one.
    payment_method_on_file: z.boolean(),
    entitlements: planEntitlementsSchema.nullable(),
    usage: billingEntitlementUsageSchema,
    // When an administrator waived this account's bill, null while nobody has.
    // Usage is still priced and shown; none of it is owed. `plan` is what the
    // account returns to when the waiver is withdrawn, and `entitlements` are
    // the waiver's rather than the plan's.
    complimentary_since: timestampSchema.nullable(),
    // Whether a change of plan is still waiting on an outcome. `plan` above says
    // what the account holds, which is not what somebody who has just pressed a
    // button is asking; a change nobody could settle is retried for hours, and
    // without this the surface offers a button that answers 409.
    plan_change_pending: z.boolean(),
  })
  .strict();
export type BillingSummary = z.infer<typeof billingSummarySchema>;

/**
 * One account as an administrator sees it: the person, and what they owe.
 *
 * `status` and `plan` are null together for an account billing has never
 * written a row for, which is one that has not signed in yet. Such an account
 * can still be waived ahead of time, so `complimentary_since` may be set on a
 * row with no plan.
 */
export const billingAccountAdminSchema = z
  .object({
    user: userSchema,
    status: z.enum(billingAccountStatuses).nullable(),
    plan: billingPlanIdSchema.nullable(),
    payment_method_on_file: z.boolean(),
    complimentary_since: timestampSchema.nullable(),
    // What this account's usage cost over the trailing window, waived or not.
    recent_cost_nanos: z.number().int().nonnegative(),
    // Where that window starts; it ends at the moment of the request.
    recent_cost_since: timestampSchema,
  })
  .strict();
export type BillingAccountAdmin = z.infer<typeof billingAccountAdminSchema>;

export const billingAccountAdminListSchema = z
  .object({
    data: z.array(billingAccountAdminSchema),
    next: z.string(),
  })
  .strict();
export type BillingAccountAdminList = z.infer<typeof billingAccountAdminListSchema>;

export const billingComplimentaryRequestSchema = z
  .object({
    complimentary: z.boolean(),
  })
  .strict();
export type BillingComplimentaryRequest = z.infer<typeof billingComplimentaryRequestSchema>;

export type { BillingPlanId } from "./pricing";

File: packages/shared/src/shared/http/billing.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import Field

from shared.billing_accounts import BillingAccountStatus
from shared.billing_credits import CreditScope
from shared.billing_plans import BillingPlanId, SubscriptionTermsVersion
from shared.credit_payments import (
    MAX_CREDIT_PURCHASE_CENTS,
    MIN_CREDIT_PURCHASE_CENTS,
    CreditPaymentStatus,
)
from shared.http.base import HttpModel
from shared.http.pricing import PlanEntitlementsResponse
from shared.http.users import UserResponse


class BillingHostedSessionRequest(HttpModel):
    """Where to send the customer back to when they are done at the provider.

    Supplied by the caller rather than configured here because the dashboard
    knows which page the person was on, and returning everyone to one fixed
    location would lose that. Both are checked against the platform's own origin
    before they are used — an open redirect on a page that follows a card being
    saved is exactly the one worth abusing.
    """

    return_url: str = Field(min_length=1, max_length=2048)
    cancel_url: str = Field(default="", max_length=2048)
    """Where to return on abandoning the page. Falls back to `return_url`, which
    is the same page and correct: a customer who did not save a card should land
    where a customer who did lands, and find out there that nothing changed."""


class BillingHostedSessionResponse(HttpModel):
    """The page to send the customer to.

    A URL and nothing else. Card details are collected, stored, and shown by the
    payment provider, and never reach this platform — which is what keeps it
    outside the scope of cardholder data rules.
    """

    url: str = Field(min_length=1, max_length=2048)


class CreditPurchaseRequest(BillingHostedSessionRequest):
    request_key: UUID
    amount_cents: int = Field(
        ge=MIN_CREDIT_PURCHASE_CENTS, le=MAX_CREDIT_PURCHASE_CENTS, strict=True
    )


class CreditPurchaseResponse(HttpModel):
    id: UUID
    amount_nanos: int = Field(gt=0)
    status: CreditPaymentStatus
    checkout_url: str | None = None
    funded_at: datetime | None = None
    reversed_nanos: int = Field(ge=0)


class CreditBalanceResponse(HttpModel):
    purchased_nanos: int
    subscription_nanos: int
    trial_nanos: int
    held_nanos: int = Field(ge=0)
    debt_nanos: int = Field(ge=0)
    available_nanos: int = Field(ge=0)


class CreditSummaryResponse(HttpModel):
    ready: bool
    compute: CreditBalanceResponse
    storage_and_transfer: CreditBalanceResponse


class UsageBudgetResponse(HttpModel):
    month_started_at: datetime
    month_ended_at: datetime
    limit_nanos: int | None = Field(ge=0)
    spent_nanos: int = Field(ge=0)
    held_nanos: int = Field(ge=0)
    available_nanos: int | None = Field(ge=0)


class BillingPlanChangeRequest(HttpModel):
    """The exact published terms selected by the customer."""

    plan: BillingPlanId
    terms_version: SubscriptionTermsVersion


class BillingPlanResponse(HttpModel):
    """Purchased subscription terms and the current billing period."""

    id: BillingPlanId
    name: str
    """What to call this plan on screen, from the same card the pricing page uses.

    Sent rather than mapped in the browser, so the account summary needs no local
    plan-name map beside the catalog it fetches."""

    terms_version: SubscriptionTermsVersion | None
    monthly_nanos: int | None = Field(ge=0)
    included_nanos: int | None = Field(ge=0)
    credit_scope: CreditScope | None
    scheduled_terms_version: SubscriptionTermsVersion | None
    scheduled_change_at: datetime | None

    period_started_at: datetime | None
    period_ended_at: datetime | None


class BillingEntitlementUsageResponse(HttpModel):
    concurrent_cpu_containers: int = Field(ge=0)
    concurrent_gpus: int = Field(ge=0)
    workspaces: int = Field(ge=0)
    members: int = Field(ge=0)
    connected_clouds: int = Field(ge=0)
    custom_domains: int = Field(ge=0)


class BillingSummaryResponse(HttpModel):
    """What this account is on and where it stands.

    Scoped to the signed-in person, who is who the provider invoices — a
    workspace does not have its own payment relationship, its owner does.
    """

    status: BillingAccountStatus
    currency: str = Field(pattern=r"^[A-Z]{3}$")
    plan: BillingPlanResponse | None = None
    """What this account is subscribed to, `None` when it is subscribed to nothing.

    Signing in puts an account on the free plan, so `None` is an account that
    never reached a billing surface — one an administrator minted — or one whose
    subscription the provider says has ended. Showing either an allowance would
    state terms nothing will hold them to, and both need the same thing next,
    which is to be put on a plan.
    """

    portal_available: bool
    """Whether there is anything at the provider for this account to manage.

    True from an account's first sign-in, which is when its customer record is
    created. False only for an account that has never signed in and so is named
    nowhere at the provider; the management page has nobody to show for one of
    those, so the dashboard offers card setup instead of a dead link.
    """

    payment_method_on_file: bool
    """Whether a card is saved, which is a different question from the one above.

    `portal_available` is true from the moment a customer record exists, so it
    cannot tell an account that has saved a card from one that has not. This can,
    and has to: how much an account is given, and whether its running work is
    stopped when that is spent, both turn on it — so a customer needs to see
    which state they are in and what changes if they attach one.
    """

    entitlements: PlanEntitlementsResponse | None = None
    """What the current plan grants, absent when the account has no plan."""

    usage: BillingEntitlementUsageResponse
    """Current account-wide consumption of every measured entitlement."""

    complimentary_since: datetime | None = None
    """When an administrator waived this account's bill, `None` while nobody has.

    Usage is still priced and shown, and nothing of it is owed. `plan` beside
    it is the subscription the account still holds and returns to when the
    waiver is withdrawn. `entitlements` are the waiver's, which are what the
    account is held to.
    """

    plan_change_pending: bool
    """Whether a change of plan is still waiting on an outcome.

    `plan` above says what the account holds, which is not what somebody who has
    just asked to move is looking for. A change nobody could settle is retried
    for hours, and a surface without this shows the old plan beside a live button
    that answers `409` — a pending change the customer cannot see is the one
    thing this field exists to prevent.
    """


class BillingAccountAdminResponse(HttpModel):
    """One account as an administrator sees it: the person, and what they owe.

    `plan` and `status` are absent together for an account billing has never
    written a row for, which is one that has not signed in yet. Such an account
    can still be waived ahead of time, which is what `complimentary_since` on a
    row with no plan means.
    """

    user: UserResponse
    status: BillingAccountStatus | None = None
    plan: BillingPlanId | None = None
    payment_method_on_file: bool = False
    complimentary_since: datetime | None = None
    recent_cost_nanos: int = Field(ge=0)
    """What this account's usage cost over the trailing window, waived or not."""
    recent_cost_since: datetime
    """Where that window starts; it ends at the moment of the request."""


class BillingAccountAdminListResponse(HttpModel):
    data: list[BillingAccountAdminResponse] = Field(default_factory=list)
    next: str = ""


class BillingComplimentaryRequest(HttpModel):
    complimentary: bool


__all__ = [
    "BillingAccountAdminListResponse",
    "BillingAccountAdminResponse",
    "BillingComplimentaryRequest",
    "BillingEntitlementUsageResponse",
    "BillingHostedSessionRequest",
    "BillingHostedSessionResponse",
    "BillingPlanChangeRequest",
    "BillingPlanResponse",
    "BillingSummaryResponse",
    "CreditBalanceResponse",
    "CreditPurchaseRequest",
    "CreditPurchaseResponse",
    "CreditSummaryResponse",
    "UsageBudgetResponse",
]
